clients/src/main/java/org/apache/kafka/common/record/BaseRecords.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.record;
+
+/**
+ * Base interface for accessing records which could be contained in the log, or an in-memory materialization of log records.
+ */
+public interface BaseRecords {
```

Comment:
More bikeshedding... One alternative we considered is to provide a shim inside `FetchResponse` which enables conversion to `Records` or `Send` as needed. For example:
```java
class SendableRecords {
  Send send();
  Records records();
}
```
For the lazy down-conversion, we don't want to expose the `Records` instance, so we can raise `UnsupportedOperationException`. I'm basically wondering if we could use this in `FetchResponse` and drop the generics. Then maybe we could get rid of the ugly casting. You may have rejected this already, but curious to hear the reasons.